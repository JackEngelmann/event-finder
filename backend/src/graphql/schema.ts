import { gql } from 'apollo-server-express'

export const typeDefs = gql`
    type Club {
        id: Int!
        name: String!
        address: String
        region: String
        contact: String
        email: String
        specials: String
        description: String
        link: String
    }

    type Event {
        admissionFee: Float
        admissionFeeWithDiscount: Float
        amountOfFloors: Int
        club: Club!
        date: String!
        description: String
        genres: [Genre]!
        id: Int!
        imageUrl: String
        link: String
        minimumAge: Int
        name: String!
        priceCategory: Int
        special: String
    }

    input CreateEventInput {
        admissionFee: Float
        admissionFeeWithDiscount: Float
        amountOfFloors: Int
        clubId: Int!
        date: String!
        description: String
        genreIds: [Int]
        image: Upload
        imageUrl: String
        link: String
        minimumAge: Int
        name: String!
        priceCategory: Int
        special: String
    }

    input UpdateEventInput {
        admissionFee: Float
        admissionFeeWithDiscount: Float
        amountOfFloors: Int
        clubId: Int!
        date: String!
        description: String
        genreIds: [Int]
        id: Int!
        image: Upload
        imageUrl: String
        link: String
        minimumAge: Int
        name: String!
        priceCategory: Int
        special: String
    }

    input CreateClubInput {
        address: String
        contact: String
        description: String
        email: String
        link: String
        name: String!
        region: String
        specials: String
    }

    input UpdateClubInput {
        address: String
        contact: String
        description: String
        email: String
        id: Int!
        link: String
        name: String!
        region: String
        specials: String
    }

    type UpdateClubPayload {
        club: Club
    }

    type CreateClubPayload {
        club: Club
    }

    type DeleteClubPayload {
        id: Int
    }

    type CreateEventPayload {
        event: Event
    }

    type UpdateEventPayload {
        event: Event
    }

    type DeleteEventPayload {
        id: Int
    }

    type Genre {
        id: Int!
        name: String!
    }

    input EventsQueryFilter {
        date: String
    }

    type Query {
        club(id: Int!): Club
        clubs: [Club!]!
        event(id: Int!): Event
        events(filter: EventsQueryFilter): [Event!]!
        genre(id: Int!): Genre
        genres: [Genre!]!
    }

    type Mutation {
        createEvent(input: CreateEventInput!): CreateEventPayload!
        updateEvent(input: UpdateEventInput!): UpdateEventPayload!
        deleteEvent(id: Int!): DeleteEventPayload!

        createClub(input: CreateClubInput!): CreateClubPayload!
        updateClub(input: UpdateClubInput!): UpdateClubPayload!
        deleteClub(id: Int!): DeleteClubPayload!
    }
`
