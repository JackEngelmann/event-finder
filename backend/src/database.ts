import sqlite3 from 'sqlite3'

sqlite3.verbose()

const clubNames = [
'Kathys Garage',
'Stolperdiele',
'LOBO',
'Kraftwerk Mitte',
'Downtown',
'Rosis',
'Altes WettbÃ¼ro',
'Standesamt ',
'Gutzkow',
'Club Mensa',
'RÃ¶sslstube (?)',
'Sektor',
'Objekt klein a',
'Blue Dance Club',
'Gisela ',
'Arteum',
'Aquarium',
'BÃ¤renzwinger (?)',
'My House',
'TBA',
'Puro ',
'Club Neu',
'Groovestation',
'Wu5 ',
'Novitatis',
'COunt down ',
'Club 11',
'TraumtÃ¤nzer',
'GAG 18',
'Borsi 34',
'HÃ¤nge Mathe',
'Scheune Dresden',
'Parkhotel Dresden',
'Paula',
'Showboxx',
'Eventwerk (?)',
'Zille',
'Chemiefabrik',
'Reithalle StraÃŸe E',
'Alter Schlachthof ',
'Puschkin',
'Tane Ju',
'Bunker (?) ',
'Beatpol',
'Ostpol',
'Der Lude',
]

const DB_SOURCE = 'db.sqlite'
const database = new sqlite3.Database(DB_SOURCE, err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    console.log('Connected to SQLite database')
    initializeClubTable()
    initializeEventTable()
})

function initializeClubTable() {
    console.log('initializeClubTable')
    const sql = `
        CREATE TABLE club (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text
        )
    `
    database.run(sql, err => {
        if (err) return // table was created before
        const sql = 'INSERT INTO club (name) VALUES (?)'
        clubNames.forEach(clubName => {
            database.run(sql, [clubName])
        })
    })
}

function initializeEventTable() {
    console.log('initializeEventTable')
    const sql = `
        CREATE TABLE event (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text,
            description text,
            date string,
            clubId integer,
            imageUrl text
        )
    `
    database.run(sql, err => {
        if (err) return // table was created before
        const sql = 'INSERT INTO event (name, description, date, clubId, imageUrl) VALUES (?, ?, ?, ?, ?)'
        database.run(sql, [
            'Highway to Hell - Halloween 2019',
`â˜  HIGHWAY TO HELL â˜ 
Traditionell am 30.10. erwarten wir Euch wieder zu unserer spektakulÃ¤ren Halloween Party Part II  :
ðŸŽƒ SPECIAL HALLOWEEN DEKO ðŸŽƒ GRUSEL TUNNEL  
ðŸŽƒANIMATION ðŸŽƒ MAKE UP MAKER 
ðŸŽƒSPECIAL HALLOWEEN DRINK 

 â˜  â˜  â˜  LINE UP & MUSIC - 2 FLOORS â˜  â˜  â˜  
Hip Hop, Charts, 90s, Electro,  Party Hits, Latin
#Louis_FOX . Red A . DJ T-vine
                  ðŸš¨ðŸš¨EINTRITTðŸš¨ðŸš¨ 
Eintritt frei bis 00:00 Uhr, danach mit KostÃ¼m 3â‚¬ ohne 5â‚¬`,
            '2019-10-24T23:00:00.000Z',
            1,
            'images/eventLobo.jpg'
        ])
        database.run(sql, [
            'ROUTINEÃ—Shape',
            `Wer dieser Tage in Form kommen will, muss Format haben. Doch der umgebende Raum ist es, in den es sich auszudehnen lohnt, so er es zulÃ¤sst - und umso mehr, wenn nicht. In vollem Umfang den Rahmen zu sprengen bedeutet, das Ganze vom Rand her zu denken. Was zunÃ¤chst wie der Umriss einer HÃ¼lle erscheint, entpuppt sich bei genauerer Betrachtung als das Wesen der Gestalt. Und wo schlieÃŸlich alles aus dem Leim geht, wird die Dimension, wird das ganze AusmaÃŸ der seelischen Katastrophe sichtbar. So erst bekommt die Sache Kontur. Routinen kÃ¶nnen hier helfen.`,
            '2019-10-24T23:00:00.000Z',
            2,
            'images/eventOka.jpg'
        ])
        database.run(sql, [
            'Rosis Teleparty - Girls Power',
`Rosis Music Television Sunday bringt die besten Musikvideos und grÃ¶ÃŸten Hits zurÃ¼ck auf die Leinwand.

Diesmal im Spotlight, die Girls der Pop & Rockmusik umrandet von weiteren groÃŸartigen Musikvideos weiterer KÃ¼nstler

EINTRITT FREI`,
            '2019-10-24T23:00:00.000Z',
            3,
            'images/eventRosis.jpg'
        ])
    })
}

export default database