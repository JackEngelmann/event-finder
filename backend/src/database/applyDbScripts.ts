import { DatabaseConfig } from '../../databaseConfig'
import { Connection } from 'typeorm'
import { AppliedScriptModel } from './entity/appliedScripts'
import { Logger } from '../logger'

const logger = new Logger()

export async function applyDbScripts(
    connection: Connection,
    databaseConfig: DatabaseConfig
) {
    await connection.query(`
        CREATE TABLE IF NOT EXISTS appliedscript (
            id INT(11) AUTO_INCREMENT PRIMARY KEY,
            name varchar(255) NOT NULL
        ) CHARSET=utf8;
    `)
    const scripts = [
        ...(databaseConfig.migrations || []),
        ...(databaseConfig.seeds || []),
    ]
    const appliedScriptModel = new AppliedScriptModel(connection)
    const appliedScripts = await appliedScriptModel.getAppliedScripts()
    const appliedScriptNames = appliedScripts.map(row => row.name)
    for (let script of scripts) {
        if (appliedScriptNames.includes(script.name)) {
            continue
        }
        logger.info(`will run db script: ${script.name}`)
        await script.up(connection)
        await appliedScriptModel.createAppliedScript({
            name: script.name,
        })
    }
}
