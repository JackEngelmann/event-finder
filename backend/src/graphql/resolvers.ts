import { IResolvers } from 'graphql-tools'
import { queryEvent } from '../queries/event'
import { queryEvents } from '../queries/events'
import { queryClubs } from '../queries/clubs'
import { queryClub } from '../queries/club'
import { queryGenre } from '../queries/genre'
import { queryGenres } from '../queries/genres'
import { queryGenresForEvent } from '../queries/genresForEvent'
import { createEvent } from '../mutations/createEvent'
import { updateEvent } from '../mutations/updateEvent'
import { deleteEvent } from '../mutations/deleteEvent'
import { createClub } from '../mutations/createClub'
import { updateClub } from '../mutations/updateClub'
import { deleteClub } from '../mutations/deleteClub'
import { AppContext } from '../appContext'
import { requireAdminPermission } from '../permissionUtils'

type Source = {
    clubId: number
    id: number
}

export const resolvers: IResolvers<Source, AppContext> = {
    Query: {
        clubs: (obj, args, appContext, info) => queryClubs(appContext),
        club: (obj, args, appContext, info) => queryClub(appContext, args.id),
        events: (obj, args, appContext, info) =>
            queryEvents(appContext, args.filter),
        event: (obj, args, appContext, info) => queryEvent(appContext, args.id),
        genre: (obj, args, appContext, info) => queryGenre(appContext, args.id),
        genres: (obj, args, appContext, info) => queryGenres(appContext),
        me: (obj, args, appContext, info) => appContext.user,
    },
    Mutation: {
        createEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            try {
                const eventId = await createEvent(appContext, args.input)
                const event = await queryEvent(appContext, eventId)
                return { event }
            } catch (err) {
                console.log(err)
            }
        },
        updateEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateEvent(appContext, args.input)
            const event = await queryEvent(appContext, args.input.id)
            return { event }
        },
        deleteEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteEvent(appContext, args.id)
            return { id: args.id }
        },
        createClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const clubId = await createClub(appContext, args.input)
            const club = await queryClub(appContext, clubId)
            return { club }
        },
        updateClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateClub(appContext, args.input)
            const club = await queryClub(appContext, args.input.id)
            return { club }
        },
        deleteClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteClub(appContext, args.id)
            return { id: args.id }
        },
    },
    Event: {
        club: (event, args, appContext, info) =>
            queryClub(appContext, event.clubId),
        genres: (event, args, appContext, info) =>
            queryGenresForEvent(appContext, event.id),
    },
}
