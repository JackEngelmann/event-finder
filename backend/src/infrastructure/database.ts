import 'reflect-metadata'
import { Connection, ConnectionOptions, createConnection, getConnectionManager } from 'typeorm'
import { ClubDataModel, ClubRepository } from '../club/orm/club'
import { EventDataModel, EventRepository } from '../event/orm/event'
import {
    EventGenreDataModel,
    EventGenreRepository,
} from '../genre/orm/eventGenre'
import { GenreDataModel, GenreRepository } from '../genre/orm/genre'
import { ImageDataModel, ImageRepository } from '../image/orm/image'
import { UserDataModel, UserRepository } from '../auth/orm/user'
import {
    AppliedScriptDataModel,
    AppliedScriptRepository,
} from '../scripts/orm/appliedScripts'
import {
    EventImageDataModel,
    EventImageRepository,
} from '../image/orm/eventImage'
import {
    ClubImageDataModel,
    ClubImageRepository,
} from '../image/orm/clubImage'
import { LinkDataModel, LinkRepository } from '../link/orm/link'
import {
    EventLinkDataModel,
    EventLinkRepository,
} from '../link/orm/eventLink'
import {
    ClubLinkDataModel,
    ClubLinkRepository,
} from '../link/orm/clubLink'
import { migration_initialSchema } from '../migrations/initialSchema'
import { migration_191219_1657_imageUrls } from '../migrations/191219-1657-imageUrls'
import { migration_200414_1824_links } from '../migrations/200414-1824-links'
import { testData } from '../seeds/testData'


export type DbScript = {
    name: string
    up(connection: Connection): Promise<any>
}

export interface DatabaseConfig {
    migrations?: DbScript[]
    seeds?: DbScript[]
    connectionOptions: ConnectionOptions
}

const migrations: DbScript[] = [
    migration_initialSchema,
    migration_191219_1657_imageUrls,
    migration_200414_1824_links,
]

export function getMigrations(): DbScript[] {
    return migrations
}

export function getSeeds(): DbScript[] {
    if (process.env.NODE_ENV === 'production') {
        return []
    }
    return [testData]
}

export async function createDbConnection(dbName?: any) {
    if (getConnectionManager().has(dbName || 'default')) {
        return
    }
    return createConnection({
        name: dbName || 'default',
        logging: ['query', 'error', 'info', 'log', 'warn'],
        entities: [
            AppliedScriptDataModel,
            ClubDataModel,
            EventDataModel,
            EventGenreDataModel,
            GenreDataModel,
            ImageDataModel,
            UserDataModel,
            EventImageDataModel,
            ClubImageDataModel,
            LinkDataModel,
            LinkRepository,
            EventLinkDataModel,
            EventLinkRepository,
            ClubLinkDataModel,
            ClubRepository,
            UserRepository,
            EventGenreRepository,
            EventRepository,
            GenreRepository,
            ClubImageRepository,
            EventImageRepository,
            ImageRepository,
            ClubLinkRepository,
            EventLinkRepository,
            AppliedScriptRepository,
        ],
        type: 'mysql',
        database: process.env.DB_NAME!,
        host: process.env.DB_HOST!,
        username: process.env.DB_USER!,
        port: parseInt(process.env.DB_PORT!, 10),
        password: process.env.DB_PASSWORD!,
        charset: 'utf8mb4',
        logger: 'file',
    })
}
