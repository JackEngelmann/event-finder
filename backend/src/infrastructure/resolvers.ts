import { IResolvers } from 'graphql-tools'
import { queryEvent } from '../event/queries/event'
import { queryEvents } from '../event/queries/events'
import { queryClubs } from '../club/queries/clubs'
import { queryClub } from '../club/queries/club'
import { queryGenre } from '../genre/queries/genre'
import { queryGenres } from '../genre/queries/genres'
import { queryGenresForEvent } from '../genre/queries/genresForEvent'
import { createEvent } from '../event/commands/createEvent'
import { updateEvent } from '../event/commands/updateEvent'
import { deleteEvent } from '../event/commands/deleteEvent'
import { createClub } from '../club/commands/createClub'
import { updateClub } from '../club/commands/updateClub'
import { deleteClub } from '../club/commands/deleteClub'
import { AppContext } from './appContext'
import { requireAdminPermission } from '../auth/authentication'
import logger from './logger'
import { uploadImage } from '../image/commands/uploadImages'
import { queryImageUrlsForEvent } from '../image/queries/imageUrlsForEvent'
import { queryImageUrlsForClub } from '../image/queries/imageUrlsForClub'
import { queryEventsForClub } from '../event/queries/eventsForClub'
import { queryLinksForEvent } from '../link/queries/linksForEvent'
import { queryLinksForClub } from '../link/queries/linksForClub'

type Source = {
    clubId: number
    id: number
}

export const resolvers: IResolvers<Source, AppContext> = {
    Query: {
        clubs: (obj, args, appContext, info) => {
            logger.info('query clubs')
            return queryClubs(appContext)
        },
        club: (obj, args, appContext, info) => {
            logger.info('query club')
            return queryClub(appContext, args.id)
        },
        events: (obj, args, appContext, info) => {
            logger.info('query events')
            return queryEvents(appContext, args.filter)
        },
        event: (obj, args, appContext, info) => {
            logger.info('query event')
            return queryEvent(appContext, args.id)
        },
        genre: (obj, args, appContext, info) => {
            logger.info('query genre')
            return queryGenre(appContext, args.id)
        },
        genres: (obj, args, appContext, info) => {
            logger.info('query genres')
            return queryGenres(appContext)
        },
        me: (obj, args, appContext, info) => {
            logger.info('query me')
            return appContext.user
        },
    },
    Mutation: {
        createEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const eventId = await createEvent(appContext, args.input)
            const event = await queryEvent(appContext, eventId)
            return { event }
        },
        updateEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateEvent(appContext, args.input)
            const event = await queryEvent(appContext, args.input.id)
            return { event }
        },
        deleteEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteEvent(appContext, args.id)
            return { id: args.id }
        },
        createClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const clubId = await createClub(appContext, args.input)
            const club = await queryClub(appContext, clubId)
            return { club }
        },
        updateClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateClub(appContext, args.input)
            const club = await queryClub(appContext, args.input.id)
            return { club }
        },
        deleteClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteClub(appContext, args.id)
            return { id: args.id }
        },
        uploadImage: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const imageUrl = await uploadImage(appContext, args.input)
            return { imageUrl }
        },
    },
    Event: {
        club: (event, args, appContext, info) => {
            return queryClub(appContext, event.clubId)
        },
        genres: (event, args, appContext, info) =>
            queryGenresForEvent(appContext, event.id),
        imageUrls: (event, args, appContext, info) =>
            queryImageUrlsForEvent(appContext, event.id),
        links: (event, args, appContext, info) =>
            queryLinksForEvent(appContext, event.id),
    },
    Club: {
        imageUrls: (club, args, appContext, info) =>
            queryImageUrlsForClub(appContext, club.id),
        events: (club, args, appContext, info) =>
            queryEventsForClub(appContext, club.id, args.fromDay),
        links: (club, args, appContext, info) =>
            queryLinksForClub(appContext, club.id),
    },
}
