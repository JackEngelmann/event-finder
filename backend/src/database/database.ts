import 'reflect-metadata'
import { databaseConfig } from '../../databaseConfig'
import { applyDbScripts } from './applyDbScripts'
import { createConnection, Connection } from 'typeorm'
import { ClubDataModel } from './entity/club'
import { EventDataModel } from './entity/event'
import { EventGenreDataModel } from './entity/eventGenre'
import { GenreDataModel } from './entity/genre'
import { ImageDataModel } from './entity/image'
import { UserDataModel } from './entity/user'
import { AppliedScriptDataModel } from './entity/appliedScripts'
import { EventImageDataModel } from './entity/eventImage'
import { ClubImageDataModel } from './entity/clubImage'

export const createDbConnection = (dbName?: any) => {
    return createConnection({
        name: dbName,
        logging: ['query', 'error', 'info', 'log', 'warn'],
        entities: [
            AppliedScriptDataModel,
            ClubDataModel,
            EventDataModel,
            EventGenreDataModel,
            GenreDataModel,
            ImageDataModel,
            UserDataModel,
            EventImageDataModel,
            ClubImageDataModel,
        ],
        ...databaseConfig.connectionOptions,
        logger: 'file',
    })
}
// .catch(err => {
//     const logger = new Logger()
//     logger.error(err)
//     throw new Error(err)
// })

export const connectionPromise = createDbConnection()
