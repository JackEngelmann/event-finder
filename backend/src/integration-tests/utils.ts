import express from 'express'
import http from 'http'
import { AppContext } from '../appContext'
import { ApolloServer } from 'apollo-server-express'
import { typeDefs } from '../graphql/schema'
import { resolvers } from '../graphql/resolvers'
import { ApolloServerTestClient, createTestClient } from 'apollo-server-testing'
import { Connection, getConnection, createConnection } from 'typeorm'
import { createDbConnection } from '../database/database'
import { databaseConfig } from '../../databaseConfig'
import { applyDbScripts } from '../database/applyDbScripts'
import { AppliedScriptDataModel } from '../database/entity/appliedScripts'
import { ClubDataModel } from '../database/entity/club'
import { EventDataModel } from '../database/entity/event'
import { EventGenreDataModel } from '../database/entity/eventGenre'
import { GenreDataModel } from '../database/entity/genre'
import { ImageDataModel } from '../database/entity/image'
import { UserDataModel } from '../database/entity/user'
import { EventImageDataModel } from '../database/entity/eventImage'
import { ClubImageDataModel } from '../database/entity/clubImage'

/**
 * GraphQL utils
 */

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name
    description
    date
    id
    imageUrls
    priceCategory
    admissionFee
    admissionFeeWithDiscount
    link
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    address
    contact
    description
    email
    id
    link
    imageUrls
    name
    region
    specials
`

export async function insertTestData(connection: Connection) {
    await connection.query(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await connection.query(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await connection.query(`
        INSERT INTO event
        (
            name,
            description,
            date,
            clubId,
            priceCategory,
            admissionFee,
            admissionFeeWithDiscount,
            special,
            minimumAge,
            amountOfFloors
        ) 
        VALUES (
            'name',
            'description',
            'date',
            1,
            2,
            10.0,
            20.0,
            'specials',
            18,
            4
        )
    `)
    await connection.query(`
        INSERT INTO genre (name) VALUES
            ('genre1'),
            ('genre2'),
            ('genre3')
    `)
    return connection.query(`
        INSERT INTO eventgenre (eventId, genreId) VALUES (1, 1)
    `)
}

/**
 * ApolloTestClient
 */

export type ApolloTestServer = {
    server: ApolloServer
    appContext: AppContext
    client: ApolloServerTestClient
    destroy: () => Promise<unknown>
}

const dbPromise = createDbConnection()

export async function createApolloTestServer(options: {
    isAdmin: boolean
    dbName: string
}) {
    return new Promise<ApolloTestServer>(async (resolve, reject) => {
        let db = await dbPromise
        const appContext: AppContext = {
            db,
            isAdmin: options.isAdmin,
        }
        await db.synchronize(true)
        await insertTestData(db)
        const server = new ApolloServer({
            typeDefs: typeDefs,
            resolvers: resolvers,
            context: appContext,
        })
        const client = createTestClient(server)

        async function destroy() {
            await server.stop()
        }

        resolve({ destroy, server, appContext, client })
    })
}

/**
 * ApolloTestHttpServer
 */

export type ApolloHttpTestServer = {
    httpServer: http.Server
    appContext: AppContext
    port: number
    server: ApolloServer
    destroy: () => Promise<unknown>
}

export async function createApolloHttpTestServer(options: {
    isAdmin: boolean
    dbName: string
}) {
    return new Promise<ApolloHttpTestServer>(async (resolve, reject) => {
        const db = await dbPromise
        await db.createQueryRunner().dropDatabase('lieblingsclubtest')
        await db.createQueryRunner().createDatabase('lieblingsclubtest')
        const appContext: AppContext = {
            db,
            isAdmin: options.isAdmin,
        }
        await insertTestData(db)
        const server = new ApolloServer({
            typeDefs: typeDefs,
            resolvers: resolvers,
            context: appContext,
        })
        const app = express()
        server.applyMiddleware({ app })
        const httpServer = await new Promise<http.Server>(resolve => {
            const l: http.Server = app.listen({ port: 0 }, () => resolve(l))
        })
        const port = (httpServer.address() as { port: number }).port

        async function destroy() {
            if (server) await server.stop()
            if (httpServer) {
                await new Promise(resolve => httpServer.close(() => resolve()))
            }
            // return await destroyTestDb(options.dbName)
        }

        resolve({ destroy, httpServer, appContext, port, server })
    })
}
