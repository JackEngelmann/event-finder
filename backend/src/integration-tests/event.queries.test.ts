import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing'
import { ApolloServer } from 'apollo-server-express'
import { typeDefs } from '../graphql/schema'
import { resolvers } from '../graphql/resolvers'
import { createTestDb, destroyTestDb } from '../database/database'
import { eventFragment, insertTestData } from './utils'

let server: ApolloServerTestClient | undefined

const DB_NAME = 'eventquerydb'

beforeEach(async done => {
    const db = await createTestDb(DB_NAME)
    await insertTestData(db)
    const newServer = new ApolloServer({
        typeDefs,
        resolvers,
        context: { db },
    })
    server = createTestClient(newServer)
    done()
})

afterEach(async done => {
    await destroyTestDb(DB_NAME)
    done()
})

describe('event queries: ', () => {
    test('events correctly', async done => {
        const result = await server!.query({
            query: `{
                events { ${eventFragment} }
            }`,
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('existing event correctly', async done => {
        const result = await server!.query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 1,
            },
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('not existing event and return undefined', async done => {
        const result = await server!.query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 2,
            },
        })
        expect(result.data!.event).toBeNull()
        done()
    })
})
