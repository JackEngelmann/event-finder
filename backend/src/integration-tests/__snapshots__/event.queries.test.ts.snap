// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event queries:  events correctly 1`] = `
Object {
  "data": Object {
    "events": Array [
      Object {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 20,
        "amountOfFloors": 4,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "date",
        "description": "description",
        "genres": Array [
          Object {
            "id": 1,
            "name": "genre1",
          },
        ],
        "id": 1,
        "imageUrls": Array [],
        "link": null,
        "minimumAge": 18,
        "name": "name",
        "priceCategory": 2,
        "special": "specials",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event queries:  existing event correctly 1`] = `
Object {
  "data": Object {
    "event": Object {
      "admissionFee": 10,
      "admissionFeeWithDiscount": 20,
      "amountOfFloors": 4,
      "club": Object {
        "id": 1,
        "name": "name",
      },
      "date": "date",
      "description": "description",
      "genres": Array [
        Object {
          "id": 1,
          "name": "genre1",
        },
      ],
      "id": 1,
      "imageUrls": Array [],
      "link": null,
      "minimumAge": 18,
      "name": "name",
      "priceCategory": 2,
      "special": "specials",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
